### DISCUSSION TITLE
Bundle size with Alpine.js

### DISCUSSION THREAD

**Discussion by vukcevicnikola** [gregsullivan/_tw#94] [General]
Hi Greg! Thanks for this awesome starter theme and things you are doing about! :)

I'm just curious about adding Alpine.js to my project with your theme, about it's size as a third-party dependency? Will it slow down loading time and general performance?  

Thanks in advance.

**Comment by gregsullivan** [comment]
Hello—Thanks for posting this! And you're very welcome—always very happy to hear people are finding _tw useful.

I use Alpine.js for almost every site I build. To get the exact answer when added within _tw, I generated a new theme, added Alpine.js (and nothing else—no plugins), and then minified and gzipped that file. The result is 15,678 bytes, or roughly 16 KB.

(Using `npm run prod` within _tw automatically creates a minified version, and gzipping of static assets is something you should confirm is active on your host.)

Alpine.js doesn't need to be render blocking, and the file created by esbuild is placed in the footer by default, so the page will be displayed without the JavaScript file blocking it.

I think 16 KB fits comfortably into most performance budgets, but I don't want to speak for everyone. In my experience, it ends up being smaller than or equal in size to a single weight / style of a web font, which seems like a very reasonable price to pay for what it provides.

**Comment by vukcevicnikola** [comment]
Hi again :)

Greg, thanks a lot for this kind of detailed answer! I found _tw super useful for custom WordPress projects and was curious just about build with Alpine.js (never used it with WP earlier).  

16kb is great job and should deliver best results, I hope you will continue to maintain this awesome project :)

**Reply by gregsullivan** [reply]
You're very welcome! I'm still using _tw for all of my client work, and I'm actively developing it with no plans to stop. I think you'll love Alpine.js, too—it's pretty unusual for me not to use it on a project at this point. I'm planning to start recording _tw screencasts in around two weeks, and I used Alpine.js to do the motion graphics, so I'm finding all sorts of uses for it!

**Reply by vukcevicnikola** [reply]
That's awesome, can't wait for those screencasts! 
Now wondering about all that code in one file, sound a little messy, but I will surely try it in one of my next projects.

And again, thanks for all these answers :)

**Reply by gregsullivan** [reply]
You're very welcome! I'll warn you that the first round of screencasts might not be that exciting: It's just going to supplement the fundamental documentation. Hopefully the experience of doing these will make it easier for me to continue making more, though!

Are you referring to all of the code from Alpine.js in one file? If you have concerns about concatenation / minification / build steps generally, I'd suggest taking a look at this post:

https://csswizardry.com/2023/10/the-three-c-concatenate-compress-cache/

**Reply by vukcevicnikola** [reply]
I'm sure your screencasts will be super useful for all of us who use _tw. 

I was referring to any .php file of our pages, where we adding tailwind classes and alpine.js together, but that's just a matter of habit I think :)  Thanks to this provided post from csswizardry!

Did you have the opportunity to work on multilingual sites with _tw and what approach do you recommend about that?

**Reply by gregsullivan** [reply]
Oh, sorry for the misunderstanding! I'll say that there are limits to how much JavaScript I'll put in a single element. Where possible, I try to keep things pretty concise, adding anything longer as a function in the `script.js` file. Sort of like this:

```js
<div x-data="{ open: false }">
    <button x-on:click="open = ! open">Toggle Dropdown</button>
 
    <div x-show="open">
        Dropdown Contents...
    </div>
</div>
```

(That example wouldn't pass accessibility requirements, but is representative of the amount of JavaScript I'm comfortable with adding that way. They're not free yet, but the headless Alpine.js components are very helpful in creating concise and accessible versions of something like the above.)

In terms of multilingual, I've generally used WPML, and not in any special ways due to _tw. Here's a language selector from a recent project, as an example:

```php
<div class="small-caps ml-2 flex items-center pb-1 text-sm font-bold leading-[1.1] tracking-[0.05em] md:ml-4 md:block md:pb-0">
	<?php do_action( 'wpml_add_language_selector' ); ?>
</div>
```

Everything else is plugin configuration, really.

**Reply by vukcevicnikola** [reply]
Thanks again for all this help! Alpine.js components are not free, but I think they are worth buying lifetime licence, they can save a lot of coding time. Also, thanks on this guidelines for WPML!

I got some project for next month and I'm sure I will be using _tw for it.
I will share it when it's done :)

**Reply by gregsullivan** [reply]
No problem at all. Good luck with your project, and please let me know how it goes!

For whatever it's worth, I paid for the Alpine.js components, and they've been worth it for me for sure.

**Reply by vukcevicnikola** [reply]
Hi Greg,

I am pleased to inform you that my project has been successfully completed using _tw. The website achieves an exceptional performance score of 100% on Google PageSpeed Insights, marking my first site to achieve this perfect score. You can see it at domain name: biro81000.me - the site has been coded based on the design provided to me by company.

**Reply by gregsullivan** [reply]
The site looks great—congratulations! (And thanks for passing along the URL!)

**Reply by vukcevicnikola** [reply]
Thanks so much for your feedback! 🙂

**Reply by gregsullivan** [reply]
You're very welcome!

**Comment by vukcevicnikola** [comment]
Hi again Greg, 

When typing my JS at javascript/script.js it doesn't work in my header file at template-parts/layout/header-content.php ? But when I type in using <script> tag, it's working well :/ 

Do you have any suggestion? thanks in advance

**Reply by gregsullivan** [reply]
Hello again! I'm not sure I totally understand what you're trying to do and what isn't working.

Are you saying code added to `script.js` isn't firing? What does the code do?

(Generally the minified version of the `script.js` file will be loaded in the footer. Just flagging that since you mention the `header-content.php` file.)

**Comment by vukcevicnikola** [comment]
Hi Greg, sorry for late reply on this topic :/ Everything works well. 

I would like to know which bundler you are using for this theme? (webpack, parcel, vite) 
Thank you in advance.

**Reply by gregsullivan** [reply]
Very happy to hear everything is now working well!

_tw uses [esbuild](https://esbuild.github.io/) as its bundler.

Please let me know if you have any other questions!

**Comment by vukcevicnikola** [comment]
Hi Greg,

I was wondering about tailwind plugins like 'autoprefixer' and 'purgecss' for removing unused classes, does the theme have it already or we need to install them? 

Also with using Alpine in our projects, do we need to edit content array in tailwind.config.js for example like this? 

content: [
  './theme/**/*.php',        // All PHP files in your theme
  './javascript/**/*.js',    // All JavaScript files (e.g., Alpine.js or custom scripts)
  './tailwind/**/*.css',     // Custom Tailwind-based CSS files, if applicable
  './**/*.html',             // Optional: HTML files, if any
],


Thank you in advance!

**Reply by gregsullivan** [reply]
> I was wondering about tailwind plugins like 'autoprefixer' and 'purgecss' for removing unused classes, does the theme have it already or we need to install them?

This is from [Tailwind's documentation](https://tailwindcss.com/docs/browser-support#vendor-prefixes):

> If you’re using the Tailwind CLI tool, vendor prefixes like this will be added automatically.

In terms of PurgeCSS, Tailwind used it historically, but no longer does because the JIT compilation approach makes it no longer necessary. Here is the key quote from the [announcement post](https://tailwindcss.com/blog/just-in-time-the-next-generation-of-tailwind-css):

> **Your CSS is identical in development and production.** Since styles are generated as they are needed, you don’t need to purge unused styles for production, which means you see the exact same CSS in all environments. Never worry about accidentally purging an important style in production again.

-----

In terms of the `content` key, you would only need to add the JavaScript folder if you were putting Tailwind classes in those files. This is conceivable if you were creating custom directives and adding, e.g., `x-bind:class` attributes there that assigned Tailwind classes. I've never personally done this, but that doesn't mean no one ever would.

You don't need to add the CSS files. They're all loaded via `tailwind.css` by the CLI and are processed directly by Tailwind.

If you have HTML files as part of your theme and you want their classes to be registered, that would work! I think this is potentially more relevant in the block theme world, but it hasn't come up for me making hybrid themes.

**Comment by vukcevicnikola** [comment]
I got it! 

Thank you for providing this part of Tailwind's documentation and fast support, now it's make sense to me.

Sincerely, Nikola.

**Reply by gregsullivan** [reply]
You're very welcome!

**Comment by vukcevicnikola** [comment]
Hi Greg, does your Algolia Search plugin use Instantsearch.js or Search by default Wordpress template file?

I'm currently building my own inspired by this one you made! Thank you in advance :)

**Reply by gregsullivan** [reply]
I'm overriding the WP Search with Algolia autocomplete template file, so it lives here: `./theme/algolia/autocomplete.php`

This isn't what I would do with unlimited time. I would prefer to use WP Search with Algolia just for its backend functionality, and then to create my own autocomplete implementation using Autocomplete or InstantSearch directly.

But I haven't had time to do that for the _tw site, and I don't expect to for a while yet. I'm pretty happy with what I have now as an interim solution.

Hope that helps—please let me know if you have any other questions!

**Comment by vukcevicnikola** [comment]
Hi Greg, 

I would like to ask, have you tried to build custom Gutenberg blocks using React with this theme and how it's went? 
If not, do you have better foundation? 

Thanks in advance!

**Reply by gregsullivan** [reply]
I don't think there's a theme that's more appropriate for use as a base for hosting custom blocks, but I'm also not sure I recommend it.

I've done it just to see how the process would go, and I didn't think it was ideal. I think you'll generally want to use a separate plugin for the custom blocks. There was a discussion thread covering [how to use filters to style custom blocks](https://github.com/gregsullivan/_tw/discussions/70) that you might find helpful.

When I included the blocks within the theme, I had a `./blocks` directory in the root that I generated using [`@wordpress/create-block`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-create-block/). It had its own `package.json` and build process, and it built into a directory in `./theme`. `./theme/functions.php` had the necessary code to include the blocks.

This worked, but it ended up feeling a bit messy to me. I could see someone else thinking otherwise, though!

**Comment by vukcevicnikola** [comment]
Hi Greg, 

Much thanks for your explanation, I did the same thing and it seemed messy to me too. 
Could you tell me your way developing with this theme, is it traditional hard cording with CPTs, or you doing other way? 

Thank you in advance!

**Reply by gregsullivan** [reply]
Hello—Sorry for the late response!

The mandate not to include CPTs in themes comes from the theme repository, where the idea is that changes in functionality—like a custom post type—should be restricted to plugins so that users can swap between themes without losing those same CPTs.

If you're building a custom theme for a client, I don't see a problem with including CPTs in the theme.

That said, ever since ACF Pro added CPT functionality, I'll sometimes just use it, as it has the added bonus of saving a JSON copy of the CPT to the theme repository if you add a `./theme/acf-json` folder.

**Comment by vukcevicnikola** [comment]
Greg, thanks one more time!

**Reply by gregsullivan** [reply]
No problem at all—happy to help!
